### PR TITLE
Add ResolvedTS spec

### PR DESCRIPTION
**Pull request by overvenus**
Signed-off-by: Neil Shen <overvenus@gmail.com>

### CODE REVIEW

File: ResolvedTS/ResolvedTS.tla [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    ELSE CHOOSE n \in s: \A m \in s: n \leq m
+
+Max(s) ==
+    IF s = {} THEN 0
+    ELSE CHOOSE n \in s: \A m \in s: n \geq m
+
+FetchTS ==
+    /\ nextTS' = nextTS + 1 \* Try not to blew up
+                            \* state space by increasing nextTS
+    /\ LET resolvedTS == IF Len(resolvedQueue) = 0 THEN 0
```
Comment (IANTHEREAL) [inline ResolvedTS/ResolvedTS.tla:137]: where to use `resolvedTS`?

File: ResolvedTS/ResolvedTS.tla [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+RTSpec == RTInit /\ [][RTNext]_vars
+
+RTSpecWithPessimistic == RTInit /\ [][RTNextWithPessimistic]_vars
+
+\* Even if there is no write, it can always output
+\* resolved ts solely by FetchTS.
+ResolveLiveness ==
+    (/\ \A rm \in RM: /\ reState[rm].minPTS = 0
+     /\ \E i \in 1..Len(chan): chan[i].type = "FetchTS" /\ i < chanOffset)
```
Comment (IANTHEREAL) [inline ResolvedTS/ResolvedTS.tla:212]: `i =< chanOffset` means the msg had been handled, right?

File: ResolvedTS/ResolvedTS.tla [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    \/ FetchTS
+    \/ HandleMsg
+
+RTSpec == RTInit /\ [][RTNext]_vars
+
+RTSpecWithPessimistic == RTInit /\ [][RTNextWithPessimistic]_vars
+
+\* Even if there is no write, it can always output
+\* resolved ts solely by FetchTS.
+ResolveLiveness ==
```
Comment (IANTHEREAL) [inline ResolvedTS/ResolvedTS.tla:210]: I think this variant is always true, and the feature you  really want to verify is that if there are any writes at some point, the `resolvedTS` would still forward,  which also means at least one state has `resolvedTS` == `fetchTS`

File: ResolvedTS/ResolvedTS.tla [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        ~> Len(resolvedQueue) # 0
+
+CommitConsistency ==
+    \A rm \in RM:
+        rmState[rm].state = "committed" =>
+            \/ (/\ rmState[rm].pTS < rmState[rm].cTS)
+            \/ (/\ rmState[rm].pTS = rmState[rm].cTS
+                /\ rmState[rm].pTS = 0)
+
+ResolvedConsistency ==
```
Comment (IANTHEREAL) [inline ResolvedTS/ResolvedTS.tla:222]: verify `resolvedQueue # << >>`

File: ResolvedTS/ResolvedTS.tla [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    /\  LET m == chan[chanOffset + 1]
+        IN  /\ reState' = [reState EXCEPT ![m.rm].minPTS = 0]
+            /\ outputTSQueue' = Append(outputTSQueue, m.ts)
+            /\ UNCHANGED <<nextTS, rmState, resolvedQueue, chan>>
+
+HandleMsgRollback ==
+    /\ CanHandleMsg("Rollback")
+    /\  LET m == chan[chanOffset + 1]
+        IN  /\ reState' = [reState EXCEPT ![m.rm].minPTS = 0]
+            /\ UNCHANGED <<nextTS, rmState, resolvedQueue, chan>>
```
Comment (IANTHEREAL) [inline ResolvedTS/ResolvedTS.tla:187]:
```suggestion
            /\ UNCHANGED <<nextTS, rmState, outputTSQueue, resolvedQueue, chan>>
```

File: ResolvedTS/ResolvedTS.tla [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    outputTSQueue, \* A queue where commit ts saves.
+    chan,          \* A channel forwards messages to a resolver.
+    chanOffset     \* An offset indicates how many messages have been handled.
+
+vars == <<rmState, reState, nextTS,
+          resolvedQueue, outputTSQueue,
+          chan, chanOffset>>
+
+Messages ==
+    [type: {"Prewrite", "Commit", "Rollback"}, rm: RM, ts: Nat]
```
Comment (IANTHEREAL) [inline ResolvedTS/ResolvedTS.tla:21]:
```suggestion
    [type: {"Prewrite", "Commit"}, rm: RM, ts: Nat]
```

File: ResolvedTS/ResolvedTS.tla [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                   msgs |-> << >>, appliedOffset |-> 0]]
+    /\  reState = [r \in RM |-> [minPTS |-> 0]]
+    /\  nextTS = 100 \* In practical, nextTS should be a nonzero value.
+    /\  resolvedQueue = << >>
+    /\  outputTSQueue = << >>
+    /\  chan = << >>
+    /\ chanOffset = 0
+
+PessimisticPrewrite(r) ==
+    LET t == nextTS - 2 \* Pessimistic Txns start TS can be stale.
```
Comment (IANTHEREAL) [inline ResolvedTS/ResolvedTS.tla:51]: don't get the point that using `nextTS-2` in this verification,  in a sense, the pessimistic transaction is exactly the same as the optimistic transaction,  for a key TS is always incremental.

File: ResolvedTS/ResolvedTS.tla [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+    /\ CanHandleMsg("Rollback")
+    /\  LET m == chan[chanOffset + 1]
+        IN  /\ reState' = [reState EXCEPT ![m.rm].minPTS = 0]
+            /\ UNCHANGED <<nextTS, rmState, resolvedQueue, chan>>
+
+HandleMsg ==
+    \/ HandleMsgPrewrite
+    \/ HandleMsgCommit
+    \/ HandleMsgFetchTS
+
```
Comment (lance6716) [inline ResolvedTS/ResolvedTS.tla:193]:
```suggestion
    \/ HandleMsgRollback
    
```

**Pull request merged**